Issue #7033: Avoid Whitebox.setInternalState which is no longer supported in OpenJDK 12

krichter722 opened the pull request:
- expose `FileContents.javadocComments` through package-protected getter in order to make code testable
- remove `MainFrameModelPowerTest.testOpenFileWithUnknownParseMode` because it tests a language construct
- simplify overly complex `TokenUtilTest.testTokenValueIncorrect2`

close #7033

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/api/FileText.java:
```diff
@@ -80,7 +80,7 @@
      * The full text contents of the file.
      */
-    private final String fullText;
+    private String fullText;
```
what about a package-private constructor instead so this can stay `final`?

krichter722 replied:
The test assigns a value. Maybe this can be handled with another constructor.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/powermock/MainFrameModelPowerTest.java:
```diff
@@ -84,28 +77,4 @@ public void testShouldAcceptFile() throws IOException {
     }
 
-    @Test
-    public void testOpenFileWithUnknownParseMode() throws CheckstyleException {
```
> remove MainFrameModelPowerTest.testOpenFileWithUnknownParseMode because it tests a language construct

I am not sure what you mean. Most likely this test is needed for code coverage. CI will let you know.

krichter722 replied:
The test checks that something happens if the `enum` would be extended. I assume: Since its state can't change after compile time the coverage checker knows that the branch can't be reached and therefore can be excluded from coverage. I'm sure that after the PMD exclusions are added, CI passes and that this is correct.

rnveach replied:
> Since its state can't change after compile time the coverage checker knows that the branch can't be reached and therefore can be excluded from coverage

Code coverage is never that smart. This is what prompted us to add stuff like this.
Code coverage is passing so I can only assume the code has changed since this test was added.

krichter722 replied:
My point is that the test doesn't test any logic. It tests that Java enums work. You want to make sure that you don't forget to handle the case after adding a value to the enum, but that's not what unit tests are for. And you'd spot it through other tools altering after a line coverage decrease.

Ususally I agree with "code coverage is never that smart", but I think here this isn't the case.

After all, the test doesn't do any harm, however I have no idea how to adapt it to OpenJDK 12 which is the goal here.

romani replied:
such tests exists only for test coverage, is code is changed and such test is not required, lets remove it.

krichter722 commented:
@rnveach Thanks for your review. Hopefully you can assist me in allowing the following PMD violations:

```
[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.api.FileContents:219 Rule:DefaultPackage Priority:3 Use explicit scoping instead of the default package private level.

[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.api.FileText:303 Rule:DefaultPackage Priority:3 Use explicit scoping instead of the default package private level.

[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.utils.TokenUtil:44 Rule:DefaultPackage Priority:3 Use explicit scoping instead of the default package private level.
```

Searching for `/* package * /` shows that you use package protected fields in the code, but I don't seem to figure out how you make them pass. Neither do I see a suppressions file configured nor a annotation or comment rule.

rnveach commented:
https://pmd.github.io/latest/pmd_rules_java_codestyle.html#defaultpackage
Seems package-private is only allowed on constructors, and not fields/methods.
Can `protected` work?

TeamCity does also not like getter that returns direct map.

> Assignment or return of field with mutable type (Errors) (1)
220: getJavadocComments() Return of Map<Integer, TextBlock> field javadocComments

krichter722 commented:
> https://pmd.github.io/latest/pmd_rules_java_codestyle.html#defaultpackage
Seems package-private is only allowed on constructors, and not fields/methods.
> Can protected work?
>
> TeamCity does also not like getter that returns direct map.

Thanks for your research :) I think that protected exposes the field more than necessary in order to write testable code. I'd rather suppress the violation because it's the right thing to do here.

rnveach commented:
@krichter722 Please fix the CI so we can continue.

krichter722 commented:
@rnveach I suppressed the PMD violations. Any idea how to satisfy TeamCity?

rnveach commented:
To satisfy TC, you must recreate the map and return that new instance instead.
See https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java#L150

rnveach commented:
@krichter722 ping

krichter722 commented:
@rnvreach I'm not sure that's a good idea. And the comment on the code 

```
    /**
     * Returns an unmodifiable map instance containing the custom messages
     * for this configuration.
     * @return unmodifiable map containing custom messages
     */
    @Override
    public Map<String, String> getMessages() {
        return new HashMap<>(messages);
    }
```

is plain wrong. It's returning a copy which avoids modification of the class' data, but it's not an unmodifiable view. Both are not solutions for this problem.

The suppression of the complaint is the way to go here because we need (protected) access to the class data and modify it. The tool which are complaining are based on the principle of suppressing warnings. They can't know whether it's right or wrong. In this case it's the right approach.

romani commented on .ci/appveyor.bat:
```diff
@@ -27,10 +27,7 @@ mvn -e verify -Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true -Dspotbugs.skip=
-!MainFrameModelPowerTest#testOpenFileWithUnknownParseMode,^
-!TokenUtilTest#testTokenValueIncorrect2,^
-!ImportControlLoaderPowerTest#testInputStreamThatFailsOnClose^
+ -Dtest=!ImportControlLoaderPowerTest#testInputStreamThatFailsOnClose^
```
Remaining will be resolved in some other issue or PR ? 
Or ?

krichter722 replied:
I'll see what I can do. Not in this one in order to keep PR small.

romani replied:
good.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/api/FileText.java:
```diff
@@ -78,14 +78,14 @@
+     * The lines of the file, without terminators.
      */
-    private final String fullText;
+    private final String[] lines;
```
If whole problem is because this field is final, let's remove final from it.

It looks very weird that we are damaging our API, in favor of some testing tool, because tool have a bug.

It is better even sacrifice coverage, but do not damage API.

Please explain selected approach, as I might miss something.

krichter722 replied:
@romani That's what is happening. `final` is removed from `fullText` and the field order is changed in order to keep final and non-final fields grouped. The diff is a bit confusing.

romani replied:
Probably I was not clear in my comment.

In fix for this issue, should be NO changes in API ( no new methods). If problem is in final modificator let's remove it in all required places. 
As last resort, we sacrifice coverage rate.

Removal of final might not like some other tool, we will suppress violations

krichter722 replied:
I don't know whether a new package-protected method can be called a change in the API. That's the purpose of the scope package-protected to allow adding such methods, make them available for classes in the same package, including tests. I don't understand your use of the word "damage" in your previous comment. What will be damaged by this change? I'd appreciate an example. I do not understand the sentence

> If problem is in final modificator let's remove it in all required places.

The modifier has to be removed in one place only.

I don't want to push this change. I'm convinced that it has to be done in order to be able to build on OpenJDK >= 12, though. I don't see any alternative for it so far, but I have to admit that I don't get your point. I hope you find the energy to explain more. Maybe I just don't get it.

romani replied:
thanks a lot for desire to understand and help to finish this PR.

Our API is not very good, full of unnecessary details and implementations and ... very long history.
We try to fix it and eventually come with pure interfaces in API package. We try our best to not create new methods in API, the less contract/api with plugins/extensions the better for us.

For protected and package scopes .... We try to practice idea that class have no friends (term from c++) at all ! That is why such scopes appear in java, friends are considered as good idea decades ago.
So class should declare his public methods and be operated only by public methods. It is very spread problem that method become "visible" for "test" code, and misused by some other code in "main" code. As result - NO friends, no exposure of non-required methods.  

That is why you can find a lot of cases in tests that we prefer reflection vs protected/package methods, and such tests are not fragile as it is usually considered. They are same ease maintainable as others. Creation of test based on reflection is also very rare case and last resort, that is explained in comment.

> Searching for /* package * / shows that you use package protected fields in the code

Not everything is good in our code base, sometime I do not have time to think about all aspects of code and discuss in PRs till code is good.

------

Issue that we try to resolve is about tests that try to use easy reflection to make 100% coverage.
If reflection approach does not work any more in jdk12 due to "final" modifiers on fields, lets remove such modifiers. We had such modifiers only because it was next to nothing to have them and it is good practice.
Lets  comment out problematic tests and see what we will get in coverage  report. We more and more become to rely on pitest for coverage, so we might sacrifice jacoco coverage partially.

You can open few more PRs to show how different versions of fix for issue.

romani commented:
TC issue, also need to be resolved, please login as guest to TC to see violation.

krichter722 commented:
@romani Thanks for you patience and your insights. I rebased and removed the final modifier from javadocComments. I don't think it's the right approach, but software development shouldn't be evangelism. You made valid points and I got some insights in C++ while reading about friends and concepts behind. I'm not sure the idea of friends applies since one has to do something purposefully wrong in order to abuse a package-protected field or method. This danger requires much less effort than having to deal with code that is unnecessarily hard to test.

rnveach commented:
https://github.com/checkstyle/checkstyle/pull/7213 was merged, please rebase to resolve conflicts.

rnveach commented:
@krichter722 What you pushed isn't compilable. You might want to run `mvn verify test` before pushing an update.

romani commented:
Better fix is done at #7247
